File: src/api/uploads.ts
/** Lab extract/confirm API client. Maps snake_case payloads to camelCase for the UI. */

import type { ExtractedMarker, TokenUsageNotice } from '../types';

export interface ExtractResult {
  extractToken: string;
  documentSha256: string;
  parserVersion: string;
  labName: string | null;
  collectedAt: string | null;
  chronologicalAge: number | null;
  markers: ExtractedMarker[];
  tokenUsage: TokenUsageNotice;
}

export class ExtractRequestError extends Error {
  readonly usage: TokenUsageNotice | null;

  constructor(message: string, usage: TokenUsageNotice | null) {
    super(message);
    this.name = 'ExtractRequestError';
    this.usage = usage;
  }
}

export interface OwnMarker {
  rawName: string;
  canonicalId: string | null;
  loincCode: string | null;
  value: number;
  unit: string;
}

export interface OwnLabResult {
  collectedAt: string | null;
  labName: string | null;
  chronologicalAge: number | null;
  confirmedAt: string;
  documentSha256: string;
  markers: OwnMarker[];
}

export interface ConfirmResult {
  labResultId: string;
  documentSha256: string;
  parserVersion: string;
  confirmedAt: string;
  markerCount: number;
}

interface ExtractedMarkerPayload {
  raw_name: string;
  canonical_id: string | null;
  loinc_code: string | null;
  value: number;
  unit: string;
  confidence: number;
  mapping_status: 'mapped' | 'unmapped';
  within_range: boolean | null;
}

interface ExtractPayload {
  extract_token: string;
  document_sha256: string;
  parser_version: string;
  lab_name: string | null;
  collected_at: string | null;
  chronological_age: number | null;
  markers: ExtractedMarkerPayload[];
  tokens_used: number;
  tokens_limit: number;
  warning: boolean;
}

interface OwnMarkerPayload {
  raw_name: string;
  canonical_id: string | null;
  loinc_code: string | null;
  value: number;
  unit: string;
}

interface OwnLabResultPayload {
  collected_at: string | null;
  lab_name: string | null;
  chronological_age: number | null;
  confirmed_at: string;
  document_sha256: string;
  markers: OwnMarkerPayload[];
}

interface OwnLabResultsPayload {
  results: OwnLabResultPayload[];
}

interface ConfirmPayload {
  lab_result_id: string;
  document_sha256: string;
  parser_version: string;
  confirmed_at: string;
  marker_count: number;
}

function apiUrl(path: string): string {
  const base = import.meta.env.VITE_API_BASE_URL ?? '';
  return `${base}${path}`;
}

function readTokenUsage(record: Record<string, unknown>, limitReached: boolean): TokenUsageNotice | null {
  if (typeof record.tokens_used !== 'number' || typeof record.tokens_limit !== 'number') {
    return null;
  }
  return {
    tokensUsed: record.tokens_used,
    tokensLimit: record.tokens_limit,
    warning: limitReached || record.warning === true,
    limitReached,
  };
}

async function readExtractFailure(response: Response): Promise<ExtractRequestError> {
  try {
    const body: unknown = await response.json();
    if (body && typeof body === 'object') {
      const record = body as Record<string, unknown>;
      const detail =
        typeof record.detail === 'string' ? record.detail : `Request failed (${response.status})`;
      return new ExtractRequestError(detail, readTokenUsage(record, response.status === 429));
    }
  } catch {
    // Fall through to status text.
  }
  return new ExtractRequestError(`Request failed (${response.status})`, null);
}

async function readError(response: Response): Promise<string> {
  try {
    const body: unknown = await response.json();
    if (body && typeof body === 'object' && 'detail' in body) {
      const detail = (body as { detail: unknown }).detail;
      if (typeof detail === 'string') {
        return detail;
      }
    }
  } catch {
    // Fall through to status text.
  }
  return `Request failed (${response.status})`;
}

function mapMarker(payload: ExtractedMarkerPayload): ExtractedMarker {
  return {
    rawName: payload.raw_name,
    canonicalId: payload.canonical_id,
    loincCode: payload.loinc_code,
    value: payload.value,
    unit: payload.unit,
    confidence: payload.confidence,
    mappingStatus: payload.mapping_status,
    withinRange: payload.within_range,
  };
}

export async function extractLabFile(token: string, file: File): Promise<ExtractResult> {
  const body = new FormData();
  body.append('file', file);
  const response = await fetch(apiUrl('/api/v1/uploads/extract'), {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
    body,
  });
  if (!response.ok) {
    throw await readExtractFailure(response);
  }
  const payload = (await response.json()) as ExtractPayload;
  return {
    extractToken: payload.extract_token,
    documentSha256: payload.document_sha256,
    parserVersion: payload.parser_version,
    labName: payload.lab_name,
    collectedAt: payload.collected_at,
    chronologicalAge: payload.chronological_age,
    markers: payload.markers.map(mapMarker),
    tokenUsage: {
      tokensUsed: payload.tokens_used,
      tokensLimit: payload.tokens_limit,
      warning: payload.warning,
    },
  };
}

export async function fetchOwnLabResults(token: string): Promise<OwnLabResult[]> {
  const response = await fetch(apiUrl('/api/v1/uploads/results'), {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!response.ok) {
    throw new Error(await readError(response));
  }
  const payload = (await response.json()) as OwnLabResultsPayload;
  return payload.results.map((panel) => ({
    collectedAt: panel.collected_at,
    labName: panel.lab_name,
    chronologicalAge: panel.chronological_age,
    confirmedAt: panel.confirmed_at,
    documentSha256: panel.document_sha256,
    markers: panel.markers.map((marker) => ({
      rawName: marker.raw_name,
      canonicalId: marker.canonical_id,
      loincCode: marker.loinc_code,
      value: marker.value,
      unit: marker.unit,
    })),
  }));
}

export async function confirmLabExtraction(
  token: string,
  input: {
    extractToken: string;
    labName: string | null;
    collectedAt: string | null;
    chronologicalAge: number | null;
    markers: { rawName: string; value: number; unit: string }[];
  }
): Promise<ConfirmResult> {
  const response = await fetch(apiUrl('/api/v1/uploads/confirm'), {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      extract_token: input.extractToken,
      lab_name: input.labName,
      collected_at: input.collectedAt,
      chronological_age: input.chronologicalAge,
      markers: input.markers.map((marker) => ({
        raw_name: marker.rawName,
        value: marker.value,
        unit: marker.unit,
      })),
    }),
  });
  if (!response.ok) {
    throw new Error(await readError(response));
  }
  const payload = (await response.json()) as ConfirmPayload;
  return {
    labResultId: payload.lab_result_id,
    documentSha256: payload.document_sha256,
    parserVersion: payload.parser_version,
    confirmedAt: payload.confirmed_at,
    markerCount: payload.marker_count,
  };
}

File: app/api/uploads.py
"""Upload extract/confirm HTTP surface."""

from __future__ import annotations

from decimal import Decimal
from typing import Annotated

from fastapi import APIRouter, Depends, File, HTTPException, Request, UploadFile, status
from fastapi.responses import JSONResponse

from app.api.accounts import get_current_user
from app.core.config import get_settings
from app.core.deps import get_upload_service
from app.core.rate_limit import resolve_client_key
from app.domain.accounts import UserRecord
from app.domain.pii import PIIValidationError
from app.domain.schemas import (
    ConfirmRequest,
    ConfirmResponse,
    ExtractedMarkerView,
    ExtractResponse,
    OwnedLabResultsResponse,
    OwnedLabResultView,
    OwnedMarkerView,
)
from app.domain.uploads import (
    EmptyPayloadError,
    ExtractSessionNotFoundError,
    GeminiBudgetExhaustedError,
    NoMarkersError,
    PayloadTooLargeError,
    RawMarker,
    UnsupportedMediaTypeError,
    UploadError,
    VisionExtractionError,
    VisionNotConfiguredError,
    usable_chronological_age,
)
from app.services.uploads import UploadService

router = APIRouter(prefix="/api/v1/uploads", tags=["uploads"])


async def gemini_budget_exhausted_handler(_request: Request, exc: Exception) -> JSONResponse:
    """Return the caller's own counter when a Gemini budget refuses the extract."""
    tokens_used = exc.tokens_used if isinstance(exc, GeminiBudgetExhaustedError) else 0
    tokens_limit = exc.tokens_limit if isinstance(exc, GeminiBudgetExhaustedError) else 1
    return JSONResponse(
        status_code=status.HTTP_429_TOO_MANY_REQUESTS,
        content={
            "detail": "Daily extraction limit reached",
            "tokens_used": tokens_used,
            "tokens_limit": tokens_limit,
            "warning": True,
        },
    )


def _http_for(exc: UploadError) -> HTTPException:
    """Map domain errors to HTTP responses without leaking file contents."""
    if isinstance(exc, UnsupportedMediaTypeError):
        return HTTPException(
            status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            detail="Unsupported file type",
        )
    if isinstance(exc, PayloadTooLargeError):
        return HTTPException(
            status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            detail="File too large",
        )
    if isinstance(exc, EmptyPayloadError):
        return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Empty file")
    if isinstance(exc, NoMarkersError):
        return HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="No numeric markers found",
        )
    if isinstance(exc, ExtractSessionNotFoundError):
        return HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Extract session expired",
        )
    if isinstance(exc, VisionNotConfiguredError):
        return HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Vision provider is not configured",
        )
    if isinstance(exc, VisionExtractionError):
        return HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Extraction failed")
    return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Upload request failed")


@router.post("/extract", response_model=ExtractResponse)
async def extract_upload(
    request: Request,
    current: Annotated[UserRecord, Depends(get_current_user)],
    upload_service: Annotated[UploadService, Depends(get_upload_service)],
    file: Annotated[UploadFile, File()],
) -> ExtractResponse:
    """Parse a PDF or image in RAM and return markers for human review."""
    settings = get_settings()
    client_key = resolve_client_key(
        real_ip=request.headers.get("x-real-ip"),
        client_host=request.client.host if request.client is not None else None,
        trust_proxy=settings.trust_proxy_headers,
    )
    payload = await file.read()
    try:
        completed = await upload_service.extract(current.id, payload, client_key=client_key)
    except GeminiBudgetExhaustedError:
        raise
    except UploadError as exc:
        raise _http_for(exc) from exc
    finally:
        del payload
    session = completed.session
    panel = session.panel
    chronological_age = usable_chronological_age(panel.chronological_age)
    chronological_age_value = float(chronological_age) if chronological_age is not None else None
    return ExtractResponse(
        extract_token=session.token,
        document_sha256=panel.document_sha256,
        parser_version=panel.parser_version,
        lab_name=panel.lab_name,
        collected_at=panel.collected_at,
        chronological_age=chronological_age_value,
        markers=[
            ExtractedMarkerView(
                raw_name=marker.raw_name,
                canonical_id=marker.canonical_id,
                loinc_code=marker.loinc_code,
                value=float(marker.value),
                unit=marker.unit,
                confidence=marker.confidence,
                mapping_status=marker.mapping_status.value,
                within_range=marker.within_range,
            )
            for marker in panel.markers
        ],
        tokens_used=completed.tokens_used,
        tokens_limit=completed.tokens_limit,
        warning=completed.warning,
    )


@router.post("/confirm", response_model=ConfirmResponse)
async def confirm_upload(
    payload: ConfirmRequest,
    current: Annotated[UserRecord, Depends(get_current_user)],
    upload_service: Annotated[UploadService, Depends(get_upload_service)],
) -> ConfirmResponse:
    """Persist reviewed values. SHA-256 comes from the extract session.

    A name or phone in the laboratory label is dropped. The numeric markers are kept.
    """
    try:
        result = await upload_service.confirm(
            current.id,
            payload.extract_token,
            lab_name=payload.lab_name,
            collected_at=payload.collected_at,
            chronological_age=(
                Decimal(str(payload.chronological_age))
                if payload.chronological_age is not None
                else None
            ),
            markers=tuple(
                RawMarker(raw_name=item.raw_name, value=item.value, unit=item.unit)
                for item in payload.markers
            ),
        )
    except PIIValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Forbidden field",
        ) from exc
    except UploadError as exc:
        raise _http_for(exc) from exc
    return ConfirmResponse(
        lab_result_id=result.lab_result_id,
        document_sha256=result.document_sha256,
        parser_version=result.parser_version,
        confirmed_at=result.confirmed_at,
        marker_count=result.marker_count,
    )


@router.get("/results", response_model=OwnedLabResultsResponse)
async def list_own_results(
    current: Annotated[UserRecord, Depends(get_current_user)],
    upload_service: Annotated[UploadService, Depends(get_upload_service)],
) -> OwnedLabResultsResponse:
    """Return panels this account confirmed. Other accounts cannot read them."""
    panels = await upload_service.list_confirmed(current.id)
    return OwnedLabResultsResponse(
        results=[
            OwnedLabResultView(
                collected_at=panel.collected_at,
                lab_name=panel.lab_name,
                chronological_age=(
                    float(panel.chronological_age) if panel.chronological_age is not None else None
                ),
                confirmed_at=panel.confirmed_at,
                document_sha256=panel.document_sha256,
                markers=[
                    OwnedMarkerView(
                        raw_name=marker.raw_name,
                        canonical_id=marker.canonical_id,
                        loinc_code=marker.loinc_code,
                        value=float(marker.value),
                        unit=marker.unit,
                    )
                    for marker in panel.markers
                ],
            )
            for panel in panels
        ]
    )
